services/src/main/java/org/keycloak/protocol/ProtocolMapperUtils.java

```diff
@@ -123,27 +123,22 @@ public static ProtocolMapperModel findLocaleMapper(KeycloakSession session) {
     }
 
 
-    public static List<Map.Entry<ProtocolMapperModel, ProtocolMapper>> getSortedProtocolMappers(KeycloakSession session, ClientSessionContext ctx) {
-        Set<ProtocolMapperModel> mapperModels = ctx.getProtocolMappers();
-        Map<ProtocolMapperModel, ProtocolMapper> result = new HashMap<>();
-
+    public static Stream<Entry<ProtocolMapperModel, ProtocolMapper>> getSortedProtocolMappers(KeycloakSession session, ClientSessionContext ctx) {
         KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
-        for (ProtocolMapperModel mapperModel : mapperModels) {
-            ProtocolMapper mapper = (ProtocolMapper) sessionFactory.getProviderFactory(ProtocolMapper.class, mapperModel.getProtocolMapper());
-            if (mapper == null) {
-                continue;
-            }
-
-            result.put(mapperModel, mapper);
-        }
-
-        return result.entrySet()
-                .stream()
-                .sorted(Comparator.comparing(ProtocolMapperUtils::compare))
-                .collect(Collectors.toList());
+        return ctx.getProtocolMappersStream()
+                .flatMap(mapperModel -> {
+                    ProtocolMapper mapper = (ProtocolMapper) sessionFactory.getProviderFactory(ProtocolMapper.class, mapperModel.getProtocolMapper());
+                    if (mapper == null)
+                        return null;
+                    Map<ProtocolMapperModel, ProtocolMapper> protocolMapperMap = new HashMap<>();
+                    protocolMapperMap.put(mapperModel, mapper);
+                    return protocolMapperMap.entrySet().stream();
```

Comment:
I wish there was a cleaner way of doing this, but that probably would require some major refactoring. Starting with Java 9 there is a static Map.entry(key, value) method that would avoid the whole creation of a HashMap just to get the single entry in it, but it is unfortunately not availble to us right now.